kernel/arch/dreamcast/include/dc/g2bus.h

```diff
@@ -31,50 +33,165 @@
+    \param  cbdata          Data to pass to the callback function.
+    \param  dir             SH4TOG2BUS or G2BUSTOSH4.
+    \param  mode            Any value; Here for compatibility.
+    \param  g2chn           See g2b_channels.
+    \param  sh4chn          Any value; Here for compatibility.
+    \retval 0               On success.
+    \retval -1              On failure. Sets errno as appropriate.
+
+    \par    Error Conditions:
+    \em     EINPROGRESS - DMA already in progress \n
+    \em     EFAULT - sh4 and/or g2bus is not 32-byte aligned \n
+    \em     EINVAL - Invalid g2chn
+    \em     EIO - I/O error
+
+*/
+int g2_dma_transfer(void *sh4, void *g2bus, size_t length, uint32_t block,
+                    g2_dma_callback_t callback, void *cbdata,
+                    uint32_t dir, uint32_t mode, uint32_t g2chn, uint32_t sh4chn);
+
+/** \brief  Initialize DMA support.
+
+    This function sets up the DMA support for transfers to/from the G2 Bus.
 
-   Update : looks like there is a formula, mode = 3 + shchn
+    \retval 0               On success (no error conditions defined).
+*/
+int g2_dma_init(void);
+
+/** \brief  Shutdown DMA support. */
+void g2_dma_shutdown(void);
+
+/** \brief  G2 context
+
+    A G2 context containing the states of IRQs and G2 DMA. This struct
+    is used in with g2_lock() and g2_unlock().
+*/
+typedef struct { 
+    uint32_t irq_state;    /** \brief IRQ state when entering a G2 critical block */
+} g2_ctx_t;
+
+/* Internal constants */
+/** \cond */ 
+#define SPU_G2DMA_SUSPEND     (*((vuint32 *)0xa05f781C))
+#define BBA_G2DMA_SUSPEND     (*((vuint32 *)0xa05f783C))
+#define CH2_G2DMA_SUSPEND     (*((vuint32 *)0xa05f785C))
+/** \endcond */
+
+/** \brief  Disable IRQs and G2 DMA
+
+    This function makes the following g2_read_*()/g2_write_*() functions atomic 
+    by disabling IRQs and G2 DMA and storing their states. Pass the context 
+    created by this function to g2_unlock() to re-enable IRQs and G2 DMA.
 
+    \return                 The context containing the IRQ and G2 DMA states.
 */
+static inline g2_ctx_t g2_lock(void) {
+    g2_ctx_t ctx;
 
-/* We use sh channel 3 here to avoid conflicts with the PVR. */
-#define SPU_DMA_MODE   6 /* should we use 6 instead, so that the formula is 3+shchn ?
-6 works too, so ... */
-#define SPU_DMA_G2CHN  0
-#define SPU_DMA_SHCHN  3
+    ctx.irq_state = irq_disable();
+
+    /* Suspend any G2 DMA */
+    SPU_G2DMA_SUSPEND = 1;
+    BBA_G2DMA_SUSPEND = 1;
+    CH2_G2DMA_SUSPEND = 1;
+
+    while(FIFO_STATUS & FIFO_SH4);
+
+    return ctx;
+}
+
+/** \brief  Enable IRQs and G2 DMA
+
+    This function restores the IRQ and G2 DMA states using the context value
+    generated by g2_lock().
+
+    \param  ctx             The context containing IRQ and G2 DMA states.
+*/
+static inline void g2_unlock(g2_ctx_t ctx) {
+    /* Restore suspended G2 DMA */
+    SPU_G2DMA_SUSPEND = 0;
+    BBA_G2DMA_SUSPEND = 0;
+    CH2_G2DMA_SUSPEND = 0;
```

Comment:
Suspend registers can be disabled. It's fine in this case write to it?

Reply:
I have tested suspending SPU DMA while it enabled but I don't think I have really tested suspending BBA DMA. Through my testing I haven't ran into a problem suspending something I am never using. Look at CH2 DMA for example. KOS doesn't use it and it's never enabled but suspending it doesn't seem to be an issue.  Is that what you mean by disabled? Suspending something that is not enabled?

Reply:
DMA suspending is switchable option:
https://github.com/DC-SWAT/DreamShell/blob/master/include/drivers/aica.h#L42
https://github.com/DC-SWAT/DreamShell/blob/master/firmware/isoldr/loader/cdda.c?ts=4#L153

Reply:
Oh I had no idea about that bit. I'll play around with it. Currently with my code changes, that bit is not set so it currently disables suspend according to your code and I haven't had issues. I'll enable it and make sure everything functions correctly.

Reply:
I added code to enable the suspend register on transfer.  By default, until you start a g2_dma_transfer, all the the channels will have suspend register disabled. I haven't encountered an issue yet.

Reply:
So it turns out that this register is not effective, it does not stop the transfer, and therefore does not limit access to the bus, which should lead to issues?